### PR TITLE
feat(sensors): introduce a GNSS time extension

### PR DESCRIPTION
**Pull request by nponsard**
# Description

This PR adds the `GnssTimeExt` extension, defining a way to send a timestamp with nanosecond precision over sensor samples.

## Testing

Testing happens in #1330 

## Changelog Entry

A sensor extension has been added, `GnssTimeExt`, that allows to read the time and date from a GNSS sensor.

## Change Checklist

- [x] I have cleaned up my [commit history][conventional-commits] and squashed fixup commits.
- [x] I have followed the [Coding Conventions][coding-conventions].
- [x] I have tested and performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.

[conventional-commits]: https://www.conventional-commits.org/en/v1.0.0/
[coding-conventions]: https://ariel-os.github.io/ariel-os/dev/docs/book/coding-conventions.html

### CODE REVIEW

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+//!   will always return nanoseconds.
+//!
+//!
+//! # For implementors
+//!
+//! Get parts with [`convert_datetime_to_parts()`].
+//! You need to return a channel with the label [`Label::OpaqueGnssTime`] containing the first
+//! part followed by a channel with the label [`Label::Opaque`] containing the second part.
+#![cfg_attr(not(context = "native"), no_std)]
+
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]:
```suggestion
#![deny(missing_docs)]

```

File: src/ariel-os-gnss-time-extension/Cargo.toml
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/Cargo.toml:1]: Can you make sure `cargo manifest-check` is happy about this manifest? We may not need to add the crate to that CI check if we land #1656 first (and to avoid unneeded conflicts).

File: src/ariel-os-gnss-time-extension/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+edition.workspace = true
+# This crate's MSRV is decoupled from Ariel OS's.
+rust-version = "1.90"
+repository.workspace = true
+
+[lints]
+workspace = true
+
+[dependencies]
+ariel-os-debug = { path = "../ariel-os-debug" }
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/Cargo.toml]: I don't think this dependency is used:

```suggestion
```

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+use ariel_os_sensors::{
+    Label, Reading as _,
+    sensor::{ReadingChannel, Sample, SampleError, Samples},
+};
+use time::{UtcDateTime, macros::utc_datetime};
+
+const ARIEL_EPOCH: UtcDateTime = utc_datetime!(2024-01-01 0:00);
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: I suppose we could make it 2025 now?

```suggestion
const ARIEL_EPOCH: UtcDateTime = utc_datetime!(2025-01-01 0:00);
```

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+//!   Some sensors may have worse than nanosecond precision but this function
+//!   will always return nanoseconds.
+//!
+//!
+//! # For implementors
+//!
+//! Get parts with [`convert_datetime_to_parts()`].
+//! You need to return a channel with the label [`Label::OpaqueGnssTime`] containing the first
+//! part followed by a channel with the label [`Label::Opaque`] containing the second part.
+#![cfg_attr(not(context = "native"), no_std)]
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: This could help with unit testing:

```suggestion
#![cfg_attr(not(any(test, context = "native")), no_std)]
```

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    Label, Reading as _,
+    sensor::{ReadingChannel, Sample, SampleError, Samples},
+};
+use time::{UtcDateTime, macros::utc_datetime};
+
+const ARIEL_EPOCH: UtcDateTime = utc_datetime!(2024-01-01 0:00);
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum GnssTimeExtError {
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: This error type should implement `core::error::Error` but that was likely not possible because `SampleError` was missing a `core::error::Error` implementation. I've opened #1684 to fix that.

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+//! - [`GnssTimeExt::time_of_fix_timestamp_nanos()`]: UTC time in nanoseconds since UNIX epoch.
+//!   Some sensors may have worse than nanosecond precision but this function
+//!   will always return nanoseconds.
+//!
+//!
+//! # For implementors
+//!
+//! Get parts with [`convert_datetime_to_parts()`].
+//! You need to return a channel with the label [`Label::OpaqueGnssTime`] containing the first
+//! part followed by a channel with the label [`Label::Opaque`] containing the second part.
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: I think it's important to insist on this as it's not obvious from the docs alone why is needed:

```suggestion
//! part *directly followed* by a channel with the label [`Label::Opaque`] containing the second part.
```

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        Ok(sample.1.value().map_err(GnssTimeExtError::Reading)?.into())
+    }
+}
+
+/// Convert time to parts to be put into Samples.
+///
+/// # Panics
+/// Panics if the seconds since Ariel's epoch overflow.
+#[must_use]
+pub fn convert_datetime_to_parts(utc_datetime: UtcDateTime) -> (i32, i32) {
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: I have quite some qualms about having a type from `time` in the public API of such a foundational crate:

- Time is complicated and time libraries tend to go extinct: there is already a couple of them in the ecosystem, including the newest `jiff` by BurnSushi.
- `time` isn't even in 1.0 and updating the dependency would be a very significant breaking change in this crate.

I unfortunately don't have an obvious fix in mind. Maybe we can accept exposing the epoch for sensor drivers (while still making sure *users* of the API don't need to work with it)?

File: src/ariel-os-gnss-time-extension/src/lib.rs (hunk 3 above)
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: Is there another way of creating this that doesn't rely on a proc-macro? I'd rather avoid these in this type of crate.

File: src/ariel-os-gnss-time-extension/src/lib.rs (hunk 7 above)
Comment (nponsard) [reply src/ariel-os-gnss-time-extension/src/lib.rs]: A solution could be to request from the sensor a timestamp in nanoseconds since UNIX epoch, subtracting the timestamp corresponding to the start of the Ariel EPOCH would work, and would also avoid using the `utc_datetime!()` proc macro.

Comment (nponsard) [reply src/ariel-os-gnss-time-extension/src/lib.rs]: implemented in commit 2dd3b7498cb945aa87882c1f924e871ab5e0965f

File: src/ariel-os-gnss-time-extension/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "ariel-os-gnss-time-extension"
+# Versioned separately from the rest of Ariel OS.
+version = "0.1.0"
+edition.workspace = true
+rust-version = "1.90"
+repository.workspace = true
+license.workspace = true
+# This crate's MSRV is decoupled from Ariel OS's.
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/Cargo.toml]: This comment should go back on the `rust-version` key.

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+/// # Panics
+/// Panics if the seconds since Ariel's epoch overflow or the nanoseconds in a second overflow,
+/// the first panic would happen if the timestamp is too far in the future or too far in the past.
+///
+/// The Ariel epoch needs to be changed before this happens.
+///
+/// For the second panic, it should never happen unless something went very wrong,
+/// as the value range should be between 1,000,000 and -1,000,000.
+#[must_use]
+pub fn convert_datetime_to_parts(utc_datetime: i64) -> (i32, i32) {
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: Could we have some unit tests of this function, with well-chosen corner cases?

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+/// Convert time to parts to be put into Samples.
+///
+/// `utc_datetime` is an UTC timestamp as nanoseconds since UNIX epoch.
+///
+/// # Panics
+/// Panics if the seconds since Ariel's epoch overflow or the nanoseconds in a second overflow,
+/// the first panic would happen if the timestamp is too far in the future or too far in the past.
+///
+/// The Ariel epoch needs to be changed before this happens.
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: Seeing this it looks to me this is somewhat similar to [`SystemTimer::elapsed()`](https://doc.rust-lang.org/stable/std/time/struct.SystemTime.html#method.elapsed) and I think this should actually return an `Option` or even an `Error`. Maybe having a dedicated error would actually allow us to keep the epoch an implementation detail (because this current text wouldn't allow me as a user to actually know when this function would or would not panic).

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    /// Returns the nanoseconds part of the time of the last fix.
+    ///
+    /// # Errors
+    ///
+    /// Returns errors if the reading is not compatible with this extension
+    /// or one [`Sample`] returned an error.
+    fn time_of_fix_delta_ns(&self) -> Result<i64, GnssTimeExtError>;
+}
+
+impl GnssTimeExt for Samples {
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: As this crate does provide an implementation for the foreign `Samples` type, could we actually seal the `GnssTimeExt` trait? I don't think there is any other type it would make sense to implement it for, is it?

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+/// `utc_datetime` is an UTC timestamp as nanoseconds since UNIX epoch.
+///
+/// # Panics
+/// Panics if the seconds since Ariel's epoch overflow or the nanoseconds in a second overflow,
+/// the first panic would happen if the timestamp is too far in the future or too far in the past.
+///
+/// The Ariel epoch needs to be changed before this happens.
+///
+/// For the second panic, it should never happen unless something went very wrong,
+/// as the value range should be between 1,000,000 and -1,000,000.
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: This would just be an actual bug in the implementation, right? I don't think this needs to be documented (unless this can be triggered by the function inputs).

You might however want to consider using the `saturing_`/`wrapping_`/`checked_` variants of the subtraction/multiplication operations, while keeping an eye on binary size.

File: src/ariel-os-gnss-time-extension/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+#![cfg_attr(not(any(test, context = "native")), no_std)]
+#![deny(missing_docs)]
+
+use ariel_os_sensors::{
+    Label, Reading as _,
+    sensor::{ReadingChannel, Sample, SampleError, Samples},
+};
+
+// 2025-01-01T00:00:00.000Z in nanoseconds
+const ARIEL_EPOCH: i64 = 1_735_689_600_000_000;
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/src/lib.rs]: How did you obtain this? (This could then go into a comment.)

File: src/ariel-os-gnss-time-extension/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "ariel-os-gnss-time-extension"
```
Comment (ROMemories) [inline src/ariel-os-gnss-time-extension/Cargo.toml]: Maybe we should name this crate that? It only makes sense with the sensor API, and `-ext` seems to be a relatively common prefix in the ecosystem already.

```suggestion
name = "ariel-os-sensors-gnss-time-ext"
```

File: src/ariel-os-gnss-time-extension/src/lib.rs (hunk 13 above)
Comment (nponsard) [reply src/ariel-os-gnss-time-extension/src/lib.rs]: I used `Deno`
The one liner is 
```
echo 'new Date("2025-01-01 0:00Z").getTime() * 1000' | deno
```

Should this be in the comment ?

Comment (ROMemories) [reply src/ariel-os-gnss-time-extension/src/lib.rs]: If there's an easy way of using coreutils's `date` for that instead, that could go in the comment.

File: src/ariel-os-gnss-time-extension/src/lib.rs (hunk 12 above)
Comment (nponsard) [reply src/ariel-os-gnss-time-extension/src/lib.rs]: Yeah that would be an implementation issue.

Comment (nponsard) [reply src/ariel-os-gnss-time-extension/src/lib.rs]: Using `checked_sub` twice adds 24B of `text` and 124B of `bss`

**Comment by nponsard** [comment]
Added unit tests, renamed the crate and fixed a few issues found with unit tests.

File: src/ariel-os-sensors-gnss-time-ext/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "ariel-os-sensors-gnss-time-ext"
+# Versioned separately from the rest of Ariel OS.
+version = "0.1.0"
+edition.workspace = true
+# This crate's MSRV is decoupled from Ariel OS's.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/Cargo.toml]:
```suggestion
# This crate's MSRV is decoupled from Ariel OS's, but should be no higher than `ariel-os-sensors`'s.
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,502 @@
+//! # GNSS Time Extension
+//!
+//! This extension allows you to read the time information from [`Samples`] produced
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: This is otherwise not mentioned anywhere:
```suggestion
//! This extension to [`ariel-os-sensors`] allows you to read the time information from [`Samples`] produced
```

File: .github/workflows/build-deploy-docs.yml [hunk 17]
```diff
@@ -66,6 +66,7 @@ jobs:
           RUSTDOCFLAGS='--cfg nightly -Z unstable-options --extern-html-root-url trouble_host=https://docs.embassy.dev/trouble-host/0.5.0 ' cargo doc \
               --no-deps \
               -p ariel-os \
+              -p ariel-os-sensors-gnss-time-ext \
```
Comment (ROMemories) [inline .github/workflows/build-deploy-docs.yml]: That's definitely a good idea and I've opened #1849 to do the same for `ariel-os-sensors`. We'll merge #1849 first if that makes more sense for referencing the items from `ariel-os-sensors` from the docs from this new crate.

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs (hunk 16 above)
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Links to `ariel-os-sensors`'s items are broken here because they are not in scope. You can specify the paths explicitly using `[`Samples`]: ariel_os_sensors::sensor::Samples`.

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,502 @@
...
+//! by a compatible GNSS driver.
+//!
+//! # For users
+//!
+//! Import the trait [`GnssTimeExt`], you then access the time information on [`Samples`] generated
+//! by a compatible driver by using those functions:
+//!
+//! - [`GnssTimeExt::time_of_fix_timestamp()`]: UTC time in seconds since UNIX epoch.
+//! - [`GnssTimeExt::time_of_fix_timestamp_nanos()`]: UTC time in nanoseconds since UNIX epoch.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]:
```suggestion
//! - [`GnssTimeExt::time_of_fix_timestamp()`]: UTC time in whole seconds since the UNIX epoch.
//! - [`GnssTimeExt::time_of_fix_timestamp_nanos()`]: UTC time in nanoseconds since the UNIX epoch.
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,502 @@
...
+mod private {
+    use ariel_os_sensors::sensor::Samples;
+
+    pub trait Sealed {}
+    impl Sealed for Samples {}
+}
+
+/// Trait to use to access time information on [`Samples`] coming from a GNSS sensor.
+pub trait GnssTimeExt: private::Sealed {
+    /// Returns the UTC time in seconds since UNIX epoch.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]:
```suggestion
    /// Returns the UTC time in whole seconds since the UNIX epoch.
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,502 @@
...
+    ///
+    /// # Errors
+    ///
+    /// Returns errors if the reading is not compatible with this extension
+    /// or one [`Sample`] returned an error.
+    fn time_of_fix_timestamp(&self) -> Result<i64, GnssTimeExtError>;
+
+    /// Returns the time of the last fix in nanoseconds since UNIX epoch.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]:
```suggestion
    /// Returns the time of the last fix in nanoseconds since the UNIX epoch.
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,502 @@
...
+        let sample = get_element_after_marker(self.samples(), Label::OpaqueGnssTime, 1)
+            .ok_or(GnssTimeExtError::InvalidSensor)?;
+
+        Ok(sample.1.value().map_err(GnssTimeExtError::Reading)?.into())
+    }
+}
+
+/// Convert time to parts to be put into Samples.
+///
+/// `timestamp_nanos` is an UTC timestamp as nanoseconds since UNIX epoch.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]:
```suggestion
/// `timestamp_nanos` is an UTC timestamp as nanoseconds since the UNIX epoch.
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,502 @@
...
+        match self {
+            GnssTimeExtError::Reading(e) => write!(f, "reading error: {e}"),
+            GnssTimeExtError::InvalidSensor => write!(f, "invalid sensor"),
+        }
+    }
+}
+
+impl core::error::Error for GnssTimeExtError {}
+
+/// Error returned when trying convert the time to two parts.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]:
```suggestion
/// Error returned when trying to convert the time to two parts.
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,502 @@
...
+    }
+}
+
+/// Convert time to parts to be put into Samples.
+///
+/// `timestamp_nanos` is an UTC timestamp as nanoseconds since UNIX epoch.
+///
+/// The first element of the tuple is to be inserted in a Channel that has the label
+/// [`Label::OpaqueGnssTime`].
+/// The second element must be inserted in a Channel right after the channel of the first element and
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: I think these "Channel"s should be `ReadingChannel`s?

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,502 @@
...
+    }
+    fn time_of_fix_delta_nanos(&self) -> Result<i64, GnssTimeExtError> {
+        let sample = get_element_after_marker(self.samples(), Label::OpaqueGnssTime, 1)
+            .ok_or(GnssTimeExtError::InvalidSensor)?;
+
+        Ok(sample.1.value().map_err(GnssTimeExtError::Reading)?.into())
+    }
+}
+
+/// Convert time to parts to be put into Samples.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]:
```suggestion
/// Convert time to parts to be put into [`Samples`].
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,502 @@
...
+
+// Only implement for Samples
+mod private {
+    use ariel_os_sensors::sensor::Samples;
+
+    pub trait Sealed {}
+    impl Sealed for Samples {}
+}
+
+/// Trait to use to access time information on [`Samples`] coming from a GNSS sensor.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]:
```suggestion
/// Trait to access time information from [`Samples`] coming from a GNSS sensor.
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 26]
```diff
@@ -0,0 +1,502 @@
...
+                + i128::from(self.time_of_fix_delta_nanos()?),
+        )
+    }
+    /// Returns the nanoseconds part of the time of the last fix.
+    ///
+    /// # Errors
+    ///
+    /// Returns errors if the reading is not compatible with this extension
+    /// or one [`Sample`] returned an error.
+    fn time_of_fix_delta_nanos(&self) -> Result<i64, GnssTimeExtError>;
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Seems like this could be [aligned with the std](https://doc.rust-lang.org/stable/std/time/struct.Duration.html#method.subsec_nanos); don't know how I missed that existing name before.
```suggestion
    fn time_of_fix_subsec_nanos(&self) -> Result<i64, GnssTimeExtError>;
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 27]
```diff
@@ -0,0 +1,502 @@
...
+    let mut peekable = iter.peekable();
+    while let Some((c, _)) = peekable.peek() {
+        if c.label() == marker {
+            break;
+        }
+        peekable.next();
+    }
+    let result = peekable.nth(position);
+
+    // if its not an opaque channel we're doing something wrong
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]:
```suggestion
    // if it's not an opaque channel we're doing something wrong
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 28]
```diff
@@ -0,0 +1,502 @@
...
+
+/// Convert time to parts to be put into Samples.
+///
+/// `timestamp_nanos` is an UTC timestamp as nanoseconds since UNIX epoch.
+///
+/// The first element of the tuple is to be inserted in a Channel that has the label
+/// [`Label::OpaqueGnssTime`].
+/// The second element must be inserted in a Channel right after the channel of the first element and
+/// that has the label [`Label::Opaque`].
+///
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs:138]: An example about using the parts wouldn't hurt, but it is likely better added after landing the first driver that uses this.

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs (hunk 16 above)
Comment (nponsard) [reply src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: I think the link wasn't made because `ariel_os_sensors` wasn't built.

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs (hunk 23 above)
Comment (nponsard) [reply src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Indeed, I changed it.

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs (hunk 28 above)
Comment (nponsard) [reply src/ariel-os-sensors-gnss-time-ext/src/lib.rs:138]: I added an simple example

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs (hunk 16 above)
Comment (ROMemories) [reply src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Ah they are indeed in scope, it might be because `ariel_os_sensors` is not built *prior* to generating the docs of `ariel_os_sensors_gnss_time_ext` then.

Comment (nponsard) [reply src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: I'm struggling to debug this, on my machine doing 

```
cargo +nightly doc -p ariel-os-sensors -p ariel-os-sensors-gnss-time-ext --no-deps --open        
```

Has the correct links, even after deleting the `target` directory.

**Comment by nponsard** [comment]
I can't figure out what's going wrong with the links between `ariel-os-sensors` and `ariel-os-sensors-gnss-time-ext`. I can't reproduce the issue outside of the CI.

I tried this: 

```
docker run -it debian:13
apt update
apt install rustup gcc clang git
rustup toolchain install nightly-2026-01-20 --target riscv32imac-unknown-none-elf --component rust-src --profile minimal --no-self-update
git clone https://github.com/ariel-os/ariel-os

cd ariel-os

git fetch --no-tags --force --update-head-ok 'https://github.com/ariel-os/ariel-os.git' '+647ad06fe0d6ce53351d726e4c2ebe4115058257:refs/commit/647ad06fe0d6ce53351d726e4c2ebe4115058257'

git checkout 647ad06fe0d6ce53351d726e4c2ebe4115058257

RUSTDOCFLAGS='--cfg nightly -Z unstable-options --extern-html-root-url trouble_host=https://docs.embassy.dev/trouble-host/0.5.0' cargo doc --no-deps -p ariel-os -p ariel-os-sensors -p ariel-os-sensors-gnss-time-ext -p coapcore --features "ariel-os-sensors/max-sample-min-count-12,bench, ble, coap, core-affinity, csprng, defmt, dns, executor-thread, external-interrupts, hwrng, i2c, mdns, multicast, net, no-boards, random, ariel-os-coap/doc, sensors, spi, storage, tcp, threading, uart, udp, usb, usb-hid, coapcore/_nightly_docs"


cat target/doc/ariel_os_sensors_gnss_time_ext/fn.convert_datetime_to_parts.html
```

Spot `<p>You would return <a href="../ariel_os_sensors/sensor/samples/struct.Samples.html"`, that means the link worked.

**Comment by nponsard** [comment]
The issues I'm having seem to happen at random when testing locally.

**Comment by nponsard** [comment]
Solved by invoking `cargo doc` another time

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 29]
```diff
@@ -0,0 +1,533 @@
...
+/// [`Label::OpaqueGnssTime`].
+/// The second element must be inserted in a [`ReadingChannel`] right after the [`ReadingChannel`] of the first element and
+/// that has the label [`Label::Opaque`].
+///
+/// You would return [`Samples`] like this:
+///
+///     // For demonstration purposes, In a normal implementation the Samples should be returned by
+///     // `wait_for_reading()` implemented by the Sensor.
+///     let sensor = SENSOR.init(MockSensor::new(ReadingChannels::from([
+///         ReadingChannel::new(Label::Altitude, 0, ariel_os_sensors::MeasurementUnit::Meter),
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Am I missing something or is the `Opaque` `ReadingChannel` missing (and this `Altitude` channel extraneous)?

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 30]
```diff
@@ -0,0 +1,533 @@
...
+    }
+    fn time_of_fix_subsec_nanos(&self) -> Result<i64, GnssTimeExtError> {
+        let sample = get_element_after_marker(self.samples(), Label::OpaqueGnssTime, 1)
+            .ok_or(GnssTimeExtError::InvalidSensor)?;
+
+        Ok(sample.1.value().map_err(GnssTimeExtError::Reading)?.into())
+    }
+}
+
+/// Convert time to parts to be put into [`Samples`].
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: (Can be ignored and done later if I am indeed missing something below):
```suggestion
/// Converts time to parts to be put into [`Samples`].
```

File: .github/workflows/main.yml [hunk 31]
```diff
@@ -316,6 +317,7 @@ jobs:
                 --no-deps \
                 -p ariel-os \
                 -p ariel-os-sensors \
+                -p ariel-os-sensors-gnss-time-ext \
```
Comment (ROMemories) [inline .github/workflows/main.yml:320]: It would be better to keep this aligned with `build-deploy-docs.yml`, but this is better done in a `ci-build:skip` follow-up, in which we also split out the rendering of `ariel-os-sensors` for clarity.

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs (hunk 29 above)
Comment (nponsard) [reply src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Ah, I copied from the the wrong test... 
I fixed it.

**Comment by nponsard** [comment]
@ROMemories squashed, rebased #1330 on it and it work :tada: 

```
[INFO ] Time of fix: Ok(1772640243) (gnss_sensor_reader gnss-sensor-reader/src/main.rs:63)
[INFO ] Time of fix ns : Ok(1772640243965000000) (gnss_sensor_reader gnss-sensor-reader/src/main.rs:64)
```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 32]
```diff
@@ -0,0 +1,533 @@
...
+}
+
+// To obtain a timestamp with nanoseconds (works only for positive timestamps)
+//  `date --date "2038-01-19T04:04:04.040404Z" -u +%s%N`
+//
+// For a negative timestamp, get the seconds timestamp with `date --date "1956-12-13T20:45:54.1343212Z" -u +%s`,
+// if the sub-second part is non-zero, add 1. Then multiply the seconds part by 1_000_000_000, and add
+// `date --date "1956-12-13T20:45:54.1343212Z" -u +%N` - 1_000_000_000.
+#[cfg(test)]
+mod tests {
```
Comment (nponsard) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Tests aren't run in CI, working on it.

Comment (nponsard) [reply src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Fixed.

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 33]
```diff
@@ -0,0 +1,548 @@
...
+
+    use ariel_os_sensors::sensor::ReadingChannels;
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: Would fix the fmt error:

```suggestion
    use ariel_os_sensors::sensor::ReadingChannels;

```

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs (hunk 32 above)
Comment (ROMemories) [reply src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: They do indeed show in the CI logs.

File: src/ariel-os-sensors-gnss-time-ext/src/lib.rs [hunk 34]
```diff
@@ -0,0 +1,548 @@
...
+    {
+        None
+    } else {
+        result
+    }
+}
+
+#[cfg(any(test, feature = "_test"))]
+/// Mock sensor module for tests.
+/// Has to be public for docs tests.
```
Comment (ROMemories) [inline src/ariel-os-sensors-gnss-time-ext/src/lib.rs]: I'd rather keep the `#[cfg]` attribute closer to the item, unless the attribute is about the doc tests themselves (like [`#[expect(clippy::doc_markdown)]`](https://rust-lang.github.io/rust-clippy/master/index.html?search=doc_mar#doc_markdown) would be):
```suggestion
/// Mock sensor module for tests.
/// Has to be public for docs tests.
#[cfg(any(test, feature = "_test"))]
```

**Review by ROMemories — COMMENTED** [review]
Good to go after these.
